File: hdf-diff/go/fold_events_test.go
package diff

import (
	"encoding/json"
	"testing"
	"time"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	validators "github.com/mitre/hdf-libs/hdf-validators/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// feMapBatchReasons projects batch-vocabulary reasons onto the event
// vocabulary (the same mapping the kernel applies), as a set.
func feMapBatchReasons(reasons []ChangeReason) map[hdf.EventChangeReason]bool {
	out := map[hdf.EventChangeReason]bool{}
	for _, r := range reasons {
		if mapped := eventReasonFor(r); mapped != "" {
			out[mapped] = true
		}
	}
	return out
}

func feFoldReasonSet(reasons []ChangeReason) map[hdf.EventChangeReason]bool {
	out := map[hdf.EventChangeReason]bool{}
	for _, r := range reasons {
		out[hdf.EventChangeReason(r)] = true
	}
	return out
}

// feJSON canonicalizes any value for deep comparison.
func feJSON(t *testing.T, v interface{}) string {
	t.Helper()
	raw, err := json.Marshal(v)
	require.NoError(t, err)
	return string(raw)
}

// feBatchDiffs runs the batch engine in systemDrift mode and indexes the
// law-relevant entries (excluding unchanged — definitionally event-free —
// and the batch-only identity-resolution states).
func feBatchDiffs(t *testing.T, seedBytes, nextBytes []byte) map[string]RequirementDiff {
	t.Helper()
	var seedDoc, nextDoc hdf.HDFResults
	require.NoError(t, json.Unmarshal(seedBytes, &seedDoc))
	require.NoError(t, json.Unmarshal(nextBytes, &nextDoc))

	batch, err := DiffHdf(seedDoc, []hdf.HDFResults{nextDoc}, Options{ComparisonMode: ModeSystemDrift})
	require.NoError(t, err)

	out := map[string]RequirementDiff{}
	for _, d := range batch.RequirementDiffs {
		switch d.State {
		case StateUnchanged, StateMoved, StateSplit, StateMerged:
			continue
		}
		out[d.ID] = d
	}
	return out
}

func feFoldLawCheck(t *testing.T, seedName, nextName string) {
	t.Helper()
	seed := aeLoadFixture(t, seedName)
	next := aeLoadFixture(t, nextName)
	events := aeDeriveStream(t, seed, next)

	result, err := FoldChangeEventsIntoComparison(seed, events)
	require.NoError(t, err)
	assert.Empty(t, result.Warnings, "a complete derived stream must fold cleanly")

	want := feBatchDiffs(t, seed, next)
	got := map[string]RequirementDiff{}
	for _, d := range result.Comparison.RequirementDiffs {
		got[d.ID] = d
	}

	require.Equal(t, len(want), len(got), "fold and batch must agree on the changed-key set")
	for id, w := range want {
		g, ok := got[id]
		require.True(t, ok, "key %s missing from fold output", id)
		assert.Equal(t, w.State, g.State, "%s state", id)
		assert.Equal(t, w.OldEffectiveStatus, g.OldEffectiveStatus, "%s oldEffectiveStatus", id)
		assert.Equal(t, w.NewEffectiveStatus, g.NewEffectiveStatus, "%s newEffectiveStatus", id)
		assert.Equal(t, feJSON(t, w.OldImpact), feJSON(t, g.OldImpact), "%s oldImpact", id)
		assert.Equal(t, feJSON(t, w.NewImpact), feJSON(t, g.NewImpact), "%s newImpact", id)
		assert.Equal(t, w.Title, g.Title, "%s title", id)
		assert.Equal(t, w.Baseline, g.Baseline, "%s baseline", id)
		assert.Equal(t, feJSON(t, w.FieldChanges), feJSON(t, g.FieldChanges), "%s fieldChanges", id)
		assert.Equal(t, feJSON(t, w.Before), feJSON(t, g.Before), "%s before", id)
		assert.Equal(t, feJSON(t, w.After), feJSON(t, g.After), "%s after", id)
		assert.Equal(t, feMapBatchReasons(w.ChangeReasons), feFoldReasonSet(g.ChangeReasons),
			"%s changeReasons (through the event-vocabulary mapping)", id)
	}
}

func TestFoldChangeEvents_MatchesBatchDiff_ScanPair(t *testing.T) {
	feFoldLawCheck(t, "scan-before.json", "scan-after.json")
}

func TestFoldChangeEvents_MatchesBatchDiff_OverridePair(t *testing.T) {
	feFoldLawCheck(t, "scan-before.json", "scan-with-override.json")
}

func TestFoldChangeEvents_OutputShape(t *testing.T) {
	seed := aeLoadFixture(t, "scan-before.json")
	next := aeLoadFixture(t, "scan-after.json")
	events := aeDeriveStream(t, seed, next)

	result, err := FoldChangeEventsIntoComparison(seed, events)
	require.NoError(t, err)
	c := result.Comparison

	assert.Equal(t, ModeSystemDrift, c.ComparisonMode)
	assert.Equal(t, "fixture.hdf-system.json", c.SystemRef, "systemRef lifted from the events")
	assert.Equal(t, "1.0.0", c.FormatVersion)
	assert.Equal(t, "2024-02-01T00:00:00Z", c.Timestamp, "deterministic: max event occurrence, never wall clock")
	assert.NotEmpty(t, c.Sources)
	require.NotNil(t, c.Summary)
	assert.Len(t, c.RequirementDiffs, 5)

	raw, err := json.Marshal(c)
	require.NoError(t, err)
	validation := validators.ValidateComparison(raw)
	assert.True(t, validation.Valid, "folded comparison must be schema-valid: %v", validation.Errors)
}

func TestFoldChangeEvents_FoldContractProperties(t *testing.T) {
	seed := aeLoadFixture(t, "scan-before.json")
	next := aeLoadFixture(t, "scan-after.json")
	events := aeDeriveStream(t, seed, next)

	base, err := FoldChangeEventsIntoComparison(seed, events)
	require.NoError(t, err)

	t.Run("duplicate delivery is idempotent", func(t *testing.T) {
		doubled, err := FoldChangeEventsIntoComparison(seed, append(append([]*hdf.HDFRequirementChangeEvent{}, events...), events...))
		require.NoError(t, err)
		assert.Equal(t, feJSON(t, base.Comparison), feJSON(t, doubled.Comparison))
	})

	t.Run("delivery order is irrelevant", func(t *testing.T) {
		reversed := make([]*hdf.HDFRequirementChangeEvent, len(events))
		for i, ev := range events {
			reversed[len(events)-1-i] = ev
		}
		shuffled, err := FoldChangeEventsIntoComparison(seed, reversed)
		require.NoError(t, err)
		assert.Equal(t, feJSON(t, base.Comparison), feJSON(t, shuffled.Comparison))
	})
}

func TestFoldChangeEvents_Anomalies(t *testing.T) {
	seed, e1, e2 := aeTwoStepChain(t)

	t.Run("missing link warns but the winner still lifts", func(t *testing.T) {
		result, err := FoldChangeEventsIntoComparison(seed, []*hdf.HDFRequirementChangeEvent{e2})
		require.NoError(t, err)
		require.Len(t, result.Warnings, 1)
		assert.Equal(t, "chainGap", result.Warnings[0].Kind)
		require.Len(t, result.Comparison.RequirementDiffs, 1)
		assert.Equal(t, StateRegressed, result.Comparison.RequirementDiffs[0].State)
		_ = e1
	})

	t.Run("content-bearing chain for an unknown key coerces to new with a warning", func(t *testing.T) {
		passing := cePassingReq()
		passing.ID = "SV-424242"
		in := ceInputs()
		in.RequirementID = "SV-424242"
		in.Sequence = 3
		orphan := ChangeEventFromPrevious(&KeyState{
			EffectiveStatus: "failed", EffectiveImpact: 0.5,
			Checksum: hdf.Checksum{Algorithm: hdf.Sha256, Value: ecVectorFailedHalf},
		}, &passing, nil, in)
		require.NotNil(t, orphan)

		result, err := FoldChangeEventsIntoComparison(seed, []*hdf.HDFRequirementChangeEvent{orphan})
		require.NoError(t, err)
		require.Len(t, result.Warnings, 1)
		assert.Equal(t, "chainGap", result.Warnings[0].Kind)
		require.Len(t, result.Comparison.RequirementDiffs, 1)
		d := result.Comparison.RequirementDiffs[0]
		assert.Equal(t, StateNew, d.State,
			"the comparison vocabulary requires a null before only on new; the warning records the anomaly")
		assert.Nil(t, d.Before)

		raw, err := json.Marshal(result.Comparison)
		require.NoError(t, err)
		assert.True(t, validators.ValidateComparison(raw).Valid)
	})

	t.Run("absent for an unknown key warns and emits no entry", func(t *testing.T) {
		in := ceInputs()
		in.RequirementID = "SV-999999"
		in.Sequence = 4
		ghost := ChangeEventFromPrevious(&KeyState{
			EffectiveStatus: "failed", EffectiveImpact: 0.5,
			Checksum: hdf.Checksum{Algorithm: hdf.Sha256, Value: ecVectorFailedHalf},
		}, nil, nil, in)
		require.NotNil(t, ghost)

		result, err := FoldChangeEventsIntoComparison(seed, []*hdf.HDFRequirementChangeEvent{ghost})
		require.NoError(t, err)
		require.Len(t, result.Warnings, 1)
		assert.Equal(t, "absentUnknown", result.Warnings[0].Kind)
		assert.Empty(t, result.Comparison.RequirementDiffs)
	})

	t.Run("empty batch folds to an empty valid comparison", func(t *testing.T) {
		result, err := FoldChangeEventsIntoComparison(seed, nil)
		require.NoError(t, err)
		assert.Empty(t, result.Warnings)
		assert.Empty(t, result.Comparison.RequirementDiffs)
		raw, err := json.Marshal(result.Comparison)
		require.NoError(t, err)
		assert.True(t, validators.ValidateComparison(raw).Valid)
	})
}

// feTimestampLessSeed builds a seed from the shared fixture with the
// document timestamp removed and a deliberately-past waiver on SV-001:
// active at the 2024 event occurrence, expired by any later wall clock —
// so a wall-clock expiry anchor is observable as a wrong old status.
func feTimestampLessSeed(t *testing.T) ([]byte, hdf.EvaluatedRequirement) {
	t.Helper()
	var doc map[string]interface{}
	require.NoError(t, json.Unmarshal(aeLoadFixture(t, "scan-before.json"), &doc))
	delete(doc, "timestamp")

	baselines, _ := doc["baselines"].([]interface{})
	first, _ := baselines[0].(map[string]interface{})
	reqs, _ := first["requirements"].([]interface{})
	var seedReq hdf.EvaluatedRequirement
	for i, rRaw := range reqs {
		r, _ := rRaw.(map[string]interface{})
		if r["id"] != "SV-001" {
			continue
		}
		typed, ok := typedRequirement(r)
		require.True(t, ok)
		typed.StatusOverrides = []hdf.StatusOverride{stMakeOverride(struct {
			Type      string
			Status    hdf.ResultStatus
			Reason    string
			AppliedAt time.Time
			ExpiresAt time.Time
		}{
			Type:      "waiver",
			Status:    hdf.Passed,
			AppliedAt: time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
			ExpiresAt: time.Date(2025, 6, 1, 0, 0, 0, 0, time.UTC),
		})}
		raw, err := json.Marshal(typed)
		require.NoError(t, err)
		var back map[string]interface{}
		require.NoError(t, json.Unmarshal(raw, &back))
		reqs[i] = back
		seedReq = typed
	}
	require.NotEmpty(t, seedReq.ID)
	out, err := json.Marshal(doc)
	require.NoError(t, err)
	return out, seedReq
}

func TestFoldChangeEvents_TimestampLessSeedAnchorsToEventOccurrence(t *testing.T) {
	seed, seedReq := feTimestampLessSeed(t)
	refTs := "2024-02-01T00:00:00Z"

	cs := ComputeEffectiveChecksum(seedReq, refTs)
	require.NotNil(t, cs)
	prev := &KeyState{
		EffectiveStatus: ComputeEffectiveStatus(seedReq, refTs),
		EffectiveImpact: ComputeEffectiveImpact(seedReq, refTs),
		Checksum:        *cs,
	}
	require.Equal(t, "passed", prev.EffectiveStatus, "waiver must be active at the event occurrence")

	in := ceInputs()
	in.RequirementID = "SV-001"
	in.Timestamp = time.Date(2024, 2, 1, 0, 0, 0, 0, time.UTC)
	in.ReferenceTimestamp = refTs
	in.PrevReferenceTimestamp = refTs

	t.Run("absent branch", func(t *testing.T) {
		ev := ChangeEventFromPrevious(prev, nil, &seedReq, in)
		require.NotNil(t, ev)
		result, err := FoldChangeEventsIntoComparison(seed, []*hdf.HDFRequirementChangeEvent{ev})
		require.NoError(t, err)
		assert.Empty(t, result.Warnings)
		require.Len(t, result.Comparison.RequirementDiffs, 1)
		assert.Equal(t, "passed", result.Comparison.RequirementDiffs[0].OldEffectiveStatus,
			"seed-side expiry must anchor to the event occurrence, never the wall clock")
	})

	t.Run("content branch", func(t *testing.T) {
		next := seedReq
		next.Impact = 0.6
		ev := ChangeEventFromPrevious(prev, &next, &seedReq, in)
		require.NotNil(t, ev)
		require.Equal(t, hdf.EventRequirementStateUpdated, ev.State)
		result, err := FoldChangeEventsIntoComparison(seed, []*hdf.HDFRequirementChangeEvent{ev})
		require.NoError(t, err)
		require.Len(t, result.Comparison.RequirementDiffs, 1)
		assert.Equal(t, "passed", result.Comparison.RequirementDiffs[0].OldEffectiveStatus,
			"seed-side expiry must anchor to the event occurrence, never the wall clock")
	})
}

File: hdf-diff/test/fold-events.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { validateComparison } from '@mitre/hdf-validators';
import { diffHdf } from '../src/diff.js';
import type { RequirementDiff, ChangeReason } from '../src/types.js';
import { changeEventFromPrevious, type KeyState } from '../src/change-event.js';
import { computeEffectiveChecksum, computeEffectiveImpact } from '../src/effective-checksum.js';
import { computeEffectiveStatus } from '../src/status.js';
import { foldChangeEventsIntoComparison } from '../src/fold-events.js';

const FIXTURES = join(dirname(fileURLToPath(import.meta.url)), 'fixtures');
const load = (name: string): string => readFileSync(join(FIXTURES, name), 'utf-8');

type Req = Record<string, unknown>;

function requirementsOf(doc: Req): Req[] {
  const baselines = (doc['baselines'] as Req[] | undefined) ?? [];
  return baselines.flatMap((b) => (b['requirements'] as Req[] | undefined) ?? []);
}

async function deriveStream(seedJson: string, nextJson: string): Promise<Req[]> {
  const seedDoc = JSON.parse(seedJson) as Req;
  const nextDoc = JSON.parse(nextJson) as Req;
  const refTs = (nextDoc['timestamp'] as string | undefined) ?? '2026-07-22T14:03:11Z';
  const prevRefTs = (seedDoc['timestamp'] as string | undefined) ?? refTs;

  const prevByKey = new Map<string, { state: KeyState; req: Req }>();
  for (const r of requirementsOf(seedDoc)) {
    prevByKey.set(r['id'] as string, {
      state: {
        effectiveStatus: computeEffectiveStatus(r, refTs),
        effectiveImpact: computeEffectiveImpact(r, refTs),
        checksum: await computeEffectiveChecksum(r, refTs),
      },
      req: r,
    });
  }

  let seq = 0;
  const mkInputs = (id: string) => ({
    eventId: `0190f6f2-0000-7000-8000-${String(++seq).padStart(12, '0')}`,
    source: 'inspec://fixture/scan',
    sequence: seq,
    systemRef: 'fixture.hdf-system.json',
    componentId: '6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60',
    requirementId: id,
    timestamp: refTs,
    referenceTimestamp: refTs,
    prevReferenceTimestamp: prevRefTs,
  });

  const events: Req[] = [];
  for (const r of requirementsOf(nextDoc)) {
    const id = r['id'] as string;
    const prev = prevByKey.get(id) ?? null;
    prevByKey.delete(id);
    const ev = await changeEventFromPrevious(prev?.state ?? null, r, prev?.req ?? null, mkInputs(id));
    if (ev) events.push(ev);
  }
  for (const id of [...prevByKey.keys()].sort()) {
    const prev = prevByKey.get(id)!;
    const ev = await changeEventFromPrevious(prev.state, null, prev.req, mkInputs(id));
    if (ev) events.push(ev);
  }
  return events;
}

/** The kernel's batch→event reason mapping, for law comparison. */
const EVENT_REASON: Partial<Record<ChangeReason, string>> = {
  resultChanged: 'resultChanged',
  overrideAdded: 'overrideAdded',
  overrideExpired: 'overrideExpired',
  overrideRemoved: 'overrideRemoved',
  impactChanged: 'impactChanged',
  effectiveImpactChanged: 'impactChanged',
};

function mapBatchReasons(reasons: ChangeReason[]): Set<string> {
  const out = new Set<string>();
  for (const r of reasons) {
    const mapped = EVENT_REASON[r];
    if (mapped) out.add(mapped);
  }
  return out;
}

async function foldLawCheck(seedName: string, nextName: string): Promise<void> {
  const seedJson = load(seedName);
  const nextJson = load(nextName);
  const events = await deriveStream(seedJson, nextJson);

  const { comparison, warnings } = await foldChangeEventsIntoComparison(seedJson, events);
  expect(warnings, 'a complete derived stream must fold cleanly').toEqual([]);

  // TS diffHdf routes systemDrift to diffSystems (system documents), unlike
  // Go's DiffHdf — a pre-existing cross-language batch routing difference.
  // Temporal mode uses the identical comparePair construction the law needs.
  const batch = diffHdf(JSON.parse(seedJson) as Req, JSON.parse(nextJson) as Req, {
    comparisonMode: 'temporal',
  });
  const want = new Map<string, RequirementDiff>();
  for (const d of batch.requirementDiffs) {
    if (['unchanged', 'moved', 'split', 'merged'].includes(d.state)) continue;
    want.set(d.id, d);
  }

  const got = new Map((comparison['requirementDiffs'] as RequirementDiff[]).map((d) => [d.id, d]));
  expect(got.size, 'fold and batch must agree on the changed-key set').toBe(want.size);
  for (const [id, w] of want) {
    const g = got.get(id);
    expect(g, `key ${id} missing from fold output`).toBeDefined();
    expect(g!.state, `${id} state`).toBe(w.state);
    expect(g!.oldEffectiveStatus, `${id} oldEffectiveStatus`).toBe(w.oldEffectiveStatus);
    expect(g!.newEffectiveStatus, `${id} newEffectiveStatus`).toBe(w.newEffectiveStatus);
    expect(g!.oldImpact, `${id} oldImpact`).toBe(w.oldImpact);
    expect(g!.newImpact, `${id} newImpact`).toBe(w.newImpact);
    expect(g!.title, `${id} title`).toBe(w.title);
    expect(g!.fieldChanges, `${id} fieldChanges`).toEqual(w.fieldChanges);
    expect(g!.before, `${id} before`).toEqual(w.before);
    expect(g!.after, `${id} after`).toEqual(w.after);
    expect(new Set(g!.changeReasons as string[]), `${id} changeReasons (through the event mapping)`).toEqual(
      mapBatchReasons(w.changeReasons),
    );
    // matchStrategy/matchConfidence are batch matching metadata a per-key
    // stream never has; excluded from the law by design.
  }
}

describe('foldChangeEventsIntoComparison', () => {
  it('satisfies the fold-batch law on the scan pair', async () => {
    await foldLawCheck('scan-before.json', 'scan-after.json');
  });

  it('satisfies the fold-batch law on the override pair', async () => {
    await foldLawCheck('scan-before.json', 'scan-with-override.json');
  });

  it('produces a deterministic, schema-valid systemDrift comparison', async () => {
    const seedJson = load('scan-before.json');
    const events = await deriveStream(seedJson, load('scan-after.json'));
    const { comparison } = await foldChangeEventsIntoComparison(seedJson, events);

    expect(comparison['comparisonMode']).toBe('systemDrift');
    expect(comparison['systemRef']).toBe('fixture.hdf-system.json');
    expect(comparison['formatVersion']).toBe('1.0.0');
    expect(comparison['timestamp']).toBe('2024-02-01T00:00:00Z');
    expect((comparison['requirementDiffs'] as RequirementDiff[]).length).toBe(5);

    const validation = validateComparison(comparison);
    expect(validation.valid, JSON.stringify(validation.errors)).toBe(true);
  });

  it('is idempotent and order-invariant', async () => {
    const seedJson = load('scan-before.json');
    const events = await deriveStream(seedJson, load('scan-after.json'));
    const base = await foldChangeEventsIntoComparison(seedJson, events);
    const doubled = await foldChangeEventsIntoComparison(seedJson, [...events, ...events]);
    const reversed = await foldChangeEventsIntoComparison(seedJson, [...events].reverse());
    expect(doubled.comparison).toEqual(base.comparison);
    expect(reversed.comparison).toEqual(base.comparison);
  });

  it('warns on a chain gap but still lifts the winner', async () => {
    const seedJson = load('scan-before.json');
    const events = await deriveStream(seedJson, load('scan-after.json'));
    const withoutSv1 = events.filter((e) => e['requirementId'] !== 'SV-001');
    const sv1 = events.find((e) => e['requirementId'] === 'SV-001')!;
    const after = sv1['after'] as Req;
    const followOn = await changeEventFromPrevious(
      {
        effectiveStatus: 'passed',
        effectiveImpact: 0.7,
        checksum: await computeEffectiveChecksum(after, '2024-02-01T00:00:00Z'),
      },
      { ...after, results: [{ status: 'failed', codeDesc: 'regressed again', startTime: '2024-03-01T00:00:00Z' }] },
      after,
      {
        eventId: '0190f6f2-0000-7000-8000-000000000099',
        source: 'inspec://fixture/scan',
        sequence: 99,
        systemRef: 'fixture.hdf-system.json',
        componentId: '6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60',
        requirementId: 'SV-001',
        timestamp: '2024-03-01T00:00:00Z',
        referenceTimestamp: '2024-03-01T00:00:00Z',
      },
    );
    const { comparison, warnings } = await foldChangeEventsIntoComparison(seedJson, [
      ...withoutSv1,
      followOn!,
    ]);
    const gaps = warnings.filter((w) => w.kind === 'chainGap');
    expect(gaps).toHaveLength(1);
    expect(gaps[0].requirementId).toBe('SV-001');
    const sv1Diff = (comparison['requirementDiffs'] as RequirementDiff[]).find((d) => d.id === 'SV-001');
    expect(sv1Diff?.state).toBe('regressed');
  });

  it('folds an empty batch to an empty valid comparison', async () => {
    const seedJson = load('scan-before.json');
    const { comparison, warnings } = await foldChangeEventsIntoComparison(seedJson, []);
    expect(warnings).toEqual([]);
    expect(comparison['requirementDiffs']).toEqual([]);
    const validation = validateComparison(comparison);
    expect(validation.valid, JSON.stringify(validation.errors)).toBe(true);
  });
});

describe('foldChangeEventsIntoComparison defensive edges', () => {
  const MINI_SEED = JSON.stringify({
    timestamp: '2026-07-01T00:00:00Z',
    baselines: [
      { name: 'a', requirements: [{ id: 'R1', impact: 0.5, tags: {}, descriptions: [{ label: 'default', data: 'd' }], results: [{ status: 'failed', codeDesc: 't', startTime: '2025-01-01T00:00:00Z' }] }] },
    ],
  });

  function rawEvent(overrides: Req): Req {
    return {
      eventId: '0190f6f2-0000-7000-8000-000000000601',
      source: 'inspec://edge/scan',
      sequence: 601,
      systemRef: 'edge.hdf-system.json',
      componentId: '6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60',
      requirementId: 'R-EDGE',
      timestamp: '2026-07-02T00:00:00Z',
      priorChecksum: null,
      state: 'new',
      before: null,
      after: {
        id: 'R-EDGE', impact: 0.5, tags: {},
        descriptions: [{ label: 'default', data: 'd' }],
        results: [{ status: 'passed', codeDesc: 't', startTime: '2026-07-02T00:00:00Z' }],
      },
      ...overrides,
    };
  }

  it('warns absentUnknown and emits no entry for an unknown tombstone', async () => {
    const { comparison, warnings } = await foldChangeEventsIntoComparison(MINI_SEED, [
      rawEvent({ state: 'absent', after: null, before: { effectiveStatus: 'failed', effectiveImpact: 0.5 } }),
    ]);
    expect(warnings).toHaveLength(1);
    expect(warnings[0].kind).toBe('absentUnknown');
    expect(comparison['requirementDiffs']).toEqual([]);
  });

  it('coerces an unknown-key content chain to new with a chainGap warning', async () => {
    const { comparison, warnings } = await foldChangeEventsIntoComparison(MINI_SEED, [
      rawEvent({ state: 'regressed', before: { effectiveStatus: 'passed', effectiveImpact: 0.5 }, priorChecksum: { algorithm: 'sha256', value: 'deadbeef' } }),
    ]);
    expect(warnings).toHaveLength(1);
    expect(warnings[0].kind).toBe('chainGap');
    const diffs = comparison['requirementDiffs'] as RequirementDiff[];
    expect(diffs).toHaveLength(1);
    expect(diffs[0].state).toBe('new');
    expect(diffs[0].before).toBeNull();
  });

  it('warns chainGap when the winning event carries no after payload', async () => {
    const { warnings } = await foldChangeEventsIntoComparison(MINI_SEED, [
      rawEvent({ requirementId: 'R1', state: 'updated', after: null, before: { effectiveStatus: 'failed', effectiveImpact: 0.5 } }),
    ]);
    expect(warnings.some((w) => w.kind === 'chainGap')).toBe(true);
  });

  it('emits an absent diff for a known tombstone', async () => {
    const { comparison, warnings } = await foldChangeEventsIntoComparison(MINI_SEED, [
      rawEvent({ requirementId: 'R1', state: 'absent', after: null, before: { effectiveStatus: 'failed', effectiveImpact: 0.5 }, priorChecksum: { algorithm: 'sha256', value: '704f62b2d0803438ad6b7b9bab45e2c4f350b7344135a2a7f8ef986d98669021' } }),
    ]);
    const diffs = comparison['requirementDiffs'] as RequirementDiff[];
    expect(diffs).toHaveLength(1);
    expect(diffs[0].state).toBe('absent');
    expect(diffs[0].after).toBeNull();
    expect(warnings.filter((w) => w.kind !== 'chainGap')).toEqual([]);
  });

  it('throws when the comparison timestamp is underivable', async () => {
    await expect(foldChangeEventsIntoComparison(JSON.stringify({ baselines: [] }), [])).rejects.toThrow(
      /cannot derive/,
    );
  });
});

describe('foldChangeEventsIntoComparison — timestamp-less seed', () => {
  // Deliberately-past waiver: active at the 2024 event occurrence, expired
  // by any later wall clock — so a wall-clock expiry anchor is observable
  // as a wrong seed-side status. Go parity:
  // TestFoldChangeEvents_TimestampLessSeedAnchorsToEventOccurrence.
  const refTs = '2024-02-01T00:00:00Z';

  function timestampLessSeed(): { seedJson: string; seedReq: Req } {
    const doc = JSON.parse(load('scan-before.json')) as Req;
    delete doc['timestamp'];
    const seedReq = requirementsOf(doc).find((r) => r['id'] === 'SV-001');
    if (!seedReq) throw new Error('fixture missing SV-001');
    seedReq['statusOverrides'] = [
      {
        type: 'waiver',
        status: 'passed',
        reason: 'accepted pending redesign',
        appliedBy: { type: 'simple', identifier: 'admin' },
        appliedAt: '2024-01-01T00:00:00Z',
        expiresAt: '2025-06-01T00:00:00Z',
      },
    ];
    return { seedJson: JSON.stringify(doc), seedReq };
  }

  const inputs = {
    eventId: '0190f6f2-0000-7000-8000-000000000099',
    source: 'inspec://fixture/scan',
    sequence: 1,
    systemRef: 'fixture.hdf-system.json',
    componentId: '6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60',
    requirementId: 'SV-001',
    timestamp: refTs,
    referenceTimestamp: refTs,
    prevReferenceTimestamp: refTs,
  };

  async function seedState(seedReq: Req): Promise<KeyState> {
    const state = {
      effectiveStatus: computeEffectiveStatus(seedReq, refTs),
      effectiveImpact: computeEffectiveImpact(seedReq, refTs),
      checksum: await computeEffectiveChecksum(seedReq, refTs),
    };
    expect(state.effectiveStatus, 'waiver must be active at the event occurrence').toBe('passed');
    return state;
  }

  it('absent branch anchors the seed-side status to the event occurrence', async () => {
    const { seedJson, seedReq } = timestampLessSeed();
    const ev = await changeEventFromPrevious(await seedState(seedReq), null, seedReq, inputs);
    expect(ev).not.toBeNull();

    const result = await foldChangeEventsIntoComparison(seedJson, [ev as Req]);
    expect(result.warnings).toEqual([]);
    const diffs = result.comparison['requirementDiffs'] as Req[];
    expect(diffs).toHaveLength(1);
    expect(diffs[0]?.['oldEffectiveStatus'], 'expiry must anchor to the event occurrence, never the wall clock').toBe('passed');
  });

  it('content branch anchors the seed-side status to the event occurrence', async () => {
    const { seedJson, seedReq } = timestampLessSeed();
    const next = { ...seedReq, impact: 0.6 };
    const ev = await changeEventFromPrevious(await seedState(seedReq), next, seedReq, inputs);
    expect(ev).not.toBeNull();
    expect((ev as Req)['state']).toBe('updated');

    const result = await foldChangeEventsIntoComparison(seedJson, [ev as Req]);
    const diffs = result.comparison['requirementDiffs'] as Req[];
    expect(diffs).toHaveLength(1);
    expect(diffs[0]?.['oldEffectiveStatus'], 'expiry must anchor to the event occurrence, never the wall clock').toBe('passed');
  });
});
